steps/expval.py

```diff
@@ -0,0 +1,159 @@
+
+    # 2. Create operators
+    ops = []
+    if backend.n_samples is not None:
+        # Operators for Backend/Simulator in sampling mode
+        for op in operators:
+            ops.append(IsingOperator(op))
+    else:
+        # Operators for Simulator exact mode
+        for op in operators:
+            ops.append(QubitOperator(op))
+
+    # 2.+1
+    # Activate the qubits that are measured but were not acted on
+    # By applying the identity
+    # Note: this is a temporary logic subject to be removed once supported by
+    # Orquestra
+
+    # Get the active qubits of the circuit
+    active_qubits = []
+    for instr in qc.data:
+        instruction_qubits = [qubit.index for qubit in instr[1]]
+        active_qubits.extend(instruction_qubits)
+
+    active_qubits = set(active_qubits)
+
+    # Get the qubits we'd like to measure
+    # Data for identities is not stored, need to account for empty terms
+    op_qubits = [term[0][0] for op in ops for term in op.terms if term]
+
+    need_to_activate = set(op_qubits) - active_qubits
+    if need_to_activate:
+        for qubit in need_to_activate:
+            # Apply the identity
+            qc.id(qubit)
+
+    # If there are still no instructions, apply identity to the first qubit
+    # Can happen for an empty circuit when measuring the identity operator
+    if not qc.data:
+        qc.id(qc.qubits[0])
+
+    # Convert to zquantum.core.circuit.Circuit
+    circuit = Circuit(qc)
+
+    # 3. Expval
+    results = _get_expval(backend, circuit, ops)
+
+    save_list(results, "expval.json")
+
+
+def _get_expval(backend, circuit, ops):
+    """Auxiliary function to get the expectation value of a list of operators
+    given a quantum circuit and a quantum backend.
+
+    In sampling mode, the same measurement outcomes are post-processed for each
+    operator. In exact mode, the statevector prepared by the quantum circuit is
+    simulated separately for each operator. This is required so that the
+    standard ``get_exact_expectation_values`` method of the ``QuantumBackend``
+    interface can be used.
+
+    Args:
+        backend (QuantumBackend): the Orquestra quantum backend to use
+        circuit (zquantum.core.circuit.Circuit): the circuit represented as an
+            OpenQASM 2.0 program
+        operators (list): a list of operators as ``openfermion.QubitOperator``
+            or ``openfermion.IsingOperator`` objects
+
+    Returns:
+        list: list of expectation values for each operator
+    """
+    results = []
+
+    if backend.n_samples is not None:
+        measurements = backend.run_circuit_and_measure(circuit)
+
+        # Iterating through the operators specified e.g., [IsingOperator("[Z0]
+        # + [Z1]"), IsingOperator("[Z1]")] to post-process the measurements
+        # outcomes
+        for op in ops:
+            expectation_values = measurements.get_expectation_values(op)
+            expectation_values = expectation_values_to_real(expectation_values)
```

Comment:
just out of curiosity: if ``op`` is hermitian, why is the output of ``measurements.get_expectation_values(op)`` not real?

Reply:
Good question! Went with how expectation values are computed internally for Orquestra backends (like for the [Qulacs simulator](https://github.com/zapatacomputing/qe-qulacs/blob/76c42cbf1d065e8fb94bc165ceb2b84c7327daff/src/python/qequlacs/simulator.py#L45)). This sort of operation is usually a safeguard to make sure that the type of the value returned is real (a complex value could sometimes cause problems.